community/rfcs/24-05-24-001-OPEA-Code-Structure.md

```diff
@@ -0,0 +1,66 @@
+**Author**
+
+[ftian1](https://github.com/ftian1), [lvliang-intel](https://github.com/lvliang-intel), [hshen14](https://github.com/hshen14)
+
+**Status**
+
+Under Review
+
+**Objective**
+
+Define a clear criteria and rule of adding new codes into OPEA projects.
+
+**Motivation**
+
+OPEA project consists of serveral repos, including GenAIExamples, GenAIInfra, GenAICompos, and so on. We need a clear definition on where the new code for a given feature should be put into for a consistent and well-orgnized code structure.
+
+
+**Design Proposal**
+
+The proposed code structure of GenAIInfra is:
+
+```
+GenAIInfra/
+├── kubernetes-addon/ # the folder implementing additional operational capabilities to Kubernetes applications
+├── microservices-connector/ # the folder containing the implementation of microservice connector on Kubernetes
+└── scripts/
+```
+
+The proposed code structure of GenAIExamples is:
+
+```
+GenAIExamples/
+└── ChatQnA/
+    ├── kubernetes/   # for deployment by kubernetes
+    │   ├── manifests/
+    │   ├── helm-charts/
+    │   └── microservices-connector/
+    ├── docker-composer/ # for deployment by docker-composer
+    │   └── composer.yaml
+    ├── dockerfile
```

Comment:
the docker file, python file and yaml files should be moved up into the docker directory given that is the context in which they will be used.

Reply:
Malini, thanks for the comment.

I need explain this purpose further. The root dir in `ChatQnA` is used to store mega service related stuff, including chatqna mega service python files, chatqna mega service yaml and the corresponding mega service docker file. and then the second level dir is used to support different cloud deployment technologies, like "kubernetes" is used to store the chatqna kubernetes deployment related stuff, and "docker-composer" is used to store the chatqna docker composer yaml file. 

why we have such design is because we think dockerfile is the essential stuff for on-premis and cloud deployment. user can use kubernetes or docker-composer to start their own container-runtime and run/manage the docker image got from the docker file. 

hoping I explain it clearly:)

Reply:
@ftian1 we agree on the importance of the dockerfile and other components, but these are only necessary for the docker based deployment. I believe you create the chatQnA.py file through your constructor code. Only the file/input that is the input to your tool and thus our tool ideally should be in the top chatQnA directory is our opinion.